dlls/ff/ff_player.cpp

```diff
@@ -8009,3 +8030,135 @@ void CFFPlayer::UpdateCamera( bool bUnassigned )
+		{
+			m_recentAttackers[i].totalDamage += dmg;
+			m_recentAttackers[i].timestamp = timestamp;
+			DevMsg( "CFFPlayer::AddRecentAttacker1: totalDamage = %f\n", m_recentAttackers[i].totalDamage );
+			return;
+		}
+	}
+
+	// if we didnt find a match, create & add new
+	DevMsg( "CFFPlayer::AddRecentAttacker2\n" );
+	//m_recentAttackers.AddToHead( RecentAttackerInfo( attackerIdx, dmg, timestamp, pAttacker ) );
+	m_recentAttackers.AddToTail( RecentAttackerInfo( attackerIdx, dmg, timestamp, pAttacker ) );
+}
+
+// returns assisted attacker that did most damage within the last MAX_ASSIST_TIME_MS or NULL if nothing found
+RecentAttackerInfo* CFFPlayer::GetTopKillAssister( CBasePlayer* killerToIgnore )
+{
+	// oops: world kills will come in null
+	bool killedByWorld = killerToIgnore == NULL;
+	const char* killerName = killedByWorld ? "world" : killerToIgnore->GetPlayerName( );
+	DevMsg( "CFFPlayer::GetTopKillAssister0: %s killer='%s'\n", GetPlayerName( ), killerName );
+
+	RecentAttackerInfo* ret = NULL;
+
+	for ( int i = 0; i < m_recentAttackers.Count( ); ++i )
+	{
+		CFFPlayer* pFFAssister = m_recentAttackers[i].pFFPlayer;
+		// this shouldnt happen, but you know
+		if ( pFFAssister == NULL )
+			continue;
+
+		// if this was a suicide and we are killerToIgnore, we dont have to check here
+		// because previous logic prevents adding ourselves to the assist list
+		bool isKiller = !killedByWorld && pFFAssister == killerToIgnore;
+
+#ifdef _DEBUG
+		DevMsg( "CFFPlayer::GetTopKillAssister1: assister='%s' cum dmg='%f' isKiller='%s'\n",
+			pFFAssister->GetPlayerName( ), m_recentAttackers[i].totalDamage, isKiller ? "yes" : "no" );
+#endif
+		// added simple filter: if they last dmged us more than MAX_ASSIST_TIME_MS ago, ignore
+		// if its the killer, dont report also as an assist
+		if ( gpGlobals->curtime - m_recentAttackers[i].timestamp > MAX_ASSIST_TIME_MS || isKiller )
+		{
+#ifdef _DEBUG
+			DevMsg( "CFFPlayer::GetTopKillAssister2" );
+#endif
+			continue;
+		}
+
+		if ( !ret || m_recentAttackers[i].totalDamage > ret->totalDamage )
+		{	
+			ret = &m_recentAttackers[i];
+			// consider in case of exact tie, take newest timestamp attacker here
+		}
+	}
+
+#ifdef _DEBUG
+	if ( ret ) 
+	{
+		DevMsg( "CFFPlayer::GetTopKillAssister3: top assister='%s':%f\n", ret->pFFPlayer->GetPlayerName( ), ret->totalDamage );
+	}
+	else 
+	{
+		DevMsg( "CFFPlayer::GetTopKillAssister4\n" );
+	}
+#endif
+	return ret;
+}
+
+void CFFPlayer::RemoveMeFromKillAssists( ) 
+{
+	// im disconnecting or switching teams or something, so 
+	// remove my contributions from assists. we can revisit this 
+	// if its too aggressive for eg, team switch and then having assist come through friendly fire
+
+	// if coming from spec or whatever, dont do anything because we couldnt have possibly damaged anything
+	// we are called before team is changed, so we don't have to worry about switching TO spec
+	if( GetTeamNumber() < TEAM_BLUE )
+		return;

```

Comment:
Yeah it was more of a general comment to ensure we dont get random crashes due to all this code in general rather than this one block.